src/libs/logging/src/file_rotate.rs

```diff
@@ -0,0 +1,266 @@
+        } else {
+            // Pretend the path exists without any actual checking.
+            true
+        };
+
+        if self.file.is_none() || !path_exists {
+            // If the old file was externally deleted and we attempt to open a new one before
+            // releasing the old handle, we get a Permission denied on Windows. Release the handle.
+            self.file = None;
+            let file = OpenOptions::new()
+                .create(true)
+                .write(true)
+                .truncate(self.truncate)
+                .append(!self.truncate)
+                .open(&self.path)?;
+            self.written_size = file.metadata()?.len();
+            self.file = Some(LineWriter::new(file));
+        }
+
+        Ok(())
+    }
+
+    fn rotate(&mut self) -> Result<()> {
+        let _ = self.file.take();
+
+        for i in (1..=self.rotate_keep).rev() {
+            let from = self.rotated_path(i)?;
+            let to = self.rotated_path(i + 1)?;
+            if from.exists() {
+                // Ignore the error
+                let _ = fs::rename(from, to);
+            }
+        }
+        if self.path.exists() {
+            let rotated_path = self.rotated_path(1)?;
+            fs::rename(&self.path, rotated_path)?;
+        }
+
+        let delete_path = self.rotated_path(self.rotate_keep + 1)?;
+        if delete_path.exists() {
+            fs::remove_file(delete_path)?;
+        }
+
+        self.written_size = 0;
+        self.next_reopen_check = Instant::now();
+        self.reopen_if_needed()?;
+
+        Ok(())
+    }
+
+    fn rotated_path(&self, i: usize) -> Result<PathBuf> {
+        let path = self.path.to_str().ok_or_else(|| {
+            io::Error::new(
+                io::ErrorKind::InvalidInput,
+                format!("Non UTF-8 log file path: {:?}", self.path),
+            )
+        })?;
+        Ok(PathBuf::from(format!("{}.{}", path, i)))
+    }
+}
+
+impl Write for FileRotator {
+    fn write(&mut self, buf: &[u8]) -> Result<usize> {
+        self.reopen_if_needed()?;
+        let file = match self.file.as_mut() {
+            Some(v) => v,
+            None => {
+                return Err(io::Error::new(
+                    io::ErrorKind::Other,
+                    format!("Cannot open file: {:?}", self.path),
+                ))
+            }
+        };
+
+        file.write(buf).map(|size| {
+            self.written_size += size as u64;
+            size
+        })
+    }
+
+    fn flush(&mut self) -> Result<()> {
+        if let Some(f) = self.file.as_mut() {
+            f.flush()?;
+        }
+        if self.written_size >= self.rotate_size {
+            self.rotate()?;
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_rotator_valid_path() {
+        /*
+        FileRotator::new("/proc/self").unwrap_err();
+        FileRotator::new("/proc/self/__does_not_exist__/log.txt").unwrap_err();
+         */
```

Comment:
Is this a leftover?

Reply:
It's kept here for advice, whether we should enable it or remove it:)